google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkPanel.java

```diff
@@ -70,46 +71,85 @@ public CloudSdkPanel() {
         .addDocumentListener(new DocumentAdapter() {
           @Override
           protected void textChanged(DocumentEvent event) {
-            checkSdk();
+            checkSdkInBackground();
           }
         });
 
-    checkSdk();
+    checkSdkInBackground();
   }
 
-  private void checkSdk() {
-    String path = cloudSdkDirectoryField.getText();
+  private void checkSdkInBackground() {
+    final String path = cloudSdkDirectoryField.getText();
+    ApplicationManager.getApplication().executeOnPooledThread(new CloudSdkChecker(path));
+  }
 
+  @VisibleForTesting
+  protected void checkSdk(String path) {
     if (StringUtil.isEmpty(path)) {
-      warningIcon.setVisible(true);
-      warningMessage.setVisible(true);
-      warningMessage.setText(
-          createErrorMessageWithLink(
-              GctBundle.message("appengine.cloudsdk.location.missing.message")));
-
+      String warningText = appendCloudSdkDownloadMessage(
+              GctBundle.message("appengine.cloudsdk.location.missing.message"));
+      showWarning(warningText, false /* setSdkDirectoryErrorState */);
       return;
     }
 
-    try {
-      new CloudSdk.Builder()
-          .sdkPath(Paths.get(path))
-          .build()
-          .validateCloudSdk();
-
-      cloudSdkDirectoryField.getTextField().setForeground(JBColor.black);
-      warningIcon.setVisible(false);
-      warningMessage.setVisible(false);
-    } catch (AppEngineException aee) {
-      cloudSdkDirectoryField.getTextField().setForeground(JBColor.red);
-      warningIcon.setVisible(true);
-      warningMessage.setVisible(true);
-      warningMessage.setText(
-          createErrorMessageWithLink(
-              GctBundle.message("appengine.cloudsdk.location.invalid.message")));
+    Set<CloudSdkValidationResult> validationResults =
+        CloudSdkService.getInstance().validateCloudSdk(Paths.get(path));
+
+    // if there are any validation errors, show the first one to the user
+    if (validationResults.size() > 0) {
+      CloudSdkValidationResult validationResult = validationResults.iterator().next();
+      String message;
+      switch (validationResult) {
+        case CLOUD_SDK_NOT_FOUND:
+          message = appendCloudSdkDownloadMessage(validationResult.getMessage());
+          break;
+        case CLOUD_SDK_VERSION_NOT_SUPPORTED:
```

Comment:
since this is the same is that default case I'm pretty sure you can combine them:

```
case CLOUD_SDK_VERSION_NOT_SUPPORTED:
default:
 message = ....
```